### ISSUE TITLE
'list installed' command?

### ISSUE THREAD

**Issue by eprozium** [sdkman/sdkman-cli#466]
Hi,

How to find what is installed with Sdkman?

There doesn't seem to be a `sdk list installed` commands, and just calling `sdk list` simply shows all potential SDKs to install.

Thanks.

**Comment by marc0der** [comment]
Please refrain from using Github issues for usage questions. Use the Gitter user-issues chat instead.

**Comment by eprozium** [comment]
@marc0der it's not a usage question :) .

**Comment by marc0der** [comment]
Sorry for sounding harsh, that was not the intention. The [website usage](http://sdkman.io/usage.html#current) page clearly states that the command to use here is `sdk current`.

**Comment by eprozium** [comment]
Thank you very much.  Not sure how I missed that one :( .

**Comment by mvysny** [comment]
Thanks, helped me as well. Intuitively I also went for sdk list --installed or something like that.

**Comment by btassone** [comment]
Probably a good thing to consider changing sdk list, to sdk available. 1 makes more sense, and 2 that frees you up for using sdk list to show the list of installed. sdk current isn't exactly clear on what it does in its name as sdk list would be for showing all installed. Better yet, get rid of sdk current and just add a sdk list --installed. Probably even better.

**Comment by nobleach** [comment]
While using `current` does show what I'm currently using, it does not seem to list the other SDKs I have installed. For example, I'm currently using Java 11, but it sure would be nice to know I also have OpenJDK 8u111 installed - and switching to it would not require a download.

**Comment by marc0der** [comment]
@btassone apologies for never replying on this but don't think I ever got notified about your comment.

I agree that renaming `list` to `available` would make a lot of sense. It would however be _very_ confusing for all our long-time users though.

Regarding the other suggestion of qualifying `sdk list` with a `--installed` flag, I don't think this is a good idea as it goes against the command syntax of sdkman. We don't use flags like this anywhere else. The syntax always (strictly) falls into the pattern:

`sdk <command> [version] [qualifier]`

If we introduce something better or more intuitive it should fall into the same pattern.
In other words, `sdk list installed` would fall into this pattern, but then what does `sdk list` do if we rename it to `available`?

My suggestion that we bring about such change in small steps: Leave `sdk list` to be what it currently is for the time being, then introduce the `installed` qualifier on top of the `list` command to replace `current`. wdyt?

**Comment by wemu** [comment]
I think the last suggestion sounds best. "sounds" meaning the way it reads or is spoken. so `sdk list installed` and `sdk list available` reads nice and looks intuitive. `sdk list` could stay the same for some time and then change or simply point to the other statements.

**Comment by tb582** [comment]
was this ever done? - confused me for a bit this AM before some googling lol

**Comment by marc0der** [comment]
@tb582 no, still waiting for the pull requests to come in :laughing:

**Comment by lcxw** [comment]
4 years later, still not solved？

**Comment by marc0der** [comment]
@lcxw the beauty of open source is that _you_ can contribute a solution.

**Comment by rgoldberg** [comment]
`sdk list` could continue as an alias for `sdk available`.  `sdk installed` could implement the new functionality.  I don't see a need to insert `list` after `sdk` and before `available|installed`.

**Comment by leo-from-spb** [comment]
I've installed several javas, how can I list them all with their paths? The 'sdk current' shows only the current one.

**Comment by ibethencourt** [comment]
The only way I've found is:

```
sdk list java | grep installed
```

Hope that helps

**Comment by jamius19** [comment]
Unless I'm missing something, `sdk current` only shows the currently active JDK version. 
It doesn't list *all the installed versions.*

As per @ibethencourt answer, this is the only way (AFAIK) for checking all currently installed versions, which I'd love to have in SdkMan itself!

If I'm missing something please correct me.
Have a nice day!

**Comment by marstalk** [comment]
I don't think it's an issue rather than inprovment. 
As @ibethencourt saying,  _sdk list java | grep installed_ is such a good way to solve the issue.

**Comment by aairey** [comment]
This doesn't work if you want to see all candidates their installed versions.

So I resorted to this:
```bash
➜ tree -L 2 ~/.sdkman/candidates/
/home/aairey/.sdkman/candidates/
└── gradle
    ├── 6.4
    └── current -> /home/aairey/.sdkman/candidates/gradle/6.4

3 directories, 0 files
```
This works as long as you use the default `local-path` for your candidates.

**Comment by nautilor** [comment]
added this into my `.zshrc`

```
sdkman() {
        [[ "$1" == "installed" ]] && sdk list $2 | grep installed || sdk "$@"
}
```

now i can run sdkman installed java

works great

**Comment by Abhinav1217** [comment]
Stumbled here because I was searching for something similar. I have 4 jdk's installed for different purposes. so when I wanted to switch between, for example jdk 8 to jdk 11, I always forget the exact identifier, "11.0.7.hs-adpt" So I was hoping there was some command which just list just the installed java instead of entire list which require me to scroll up and search for the correct identifier. 

I am surprised this isn't readily available since the list of sdk and jdk has gotten really large. An ideal case would be something like _sdk switch java_, which would show an _alternatives_ like menu to switch java version.

**Comment by rgoldberg** [comment]
@Abhinav1217: If you use zsh, [oh-my-zsh](https://ohmyz.sh) has command line argument completion for sdkman.

Commands that only work with installed versions will only display / autocomplete installed versions.

I don’t know if sdkman argument completion exists for bash, or how well any such argument completion works.

**Comment by Abhinav1217** [comment]
I am on primarily bash, But I have omz installed I can switch to it temporarily. Currently I set sdkman to offline to limit the list to installed only.

**Comment by barmalei** [comment]
Hi Guys from 2020. Is there a SDKMAN command that list all locally installed packages or it is better to back to the question in 2021 or 2022 ?

**Comment by Abhinav1217** [comment]
@barmalei Not yet. Best I usually work with is set sdkman to offline and then list Java, It shows only installed versions, making it easier to switch. I did create a script for sdkman to do it automatically, but I usually ends up typing whole set as I distro-hop quite frequently.

**Comment by gaochundong** [comment]
+1

**Comment by vijayprasath99** [comment]
a must have feature

**Comment by antonmry** [comment]
I use shell autocompletion with `sdk use` and it works well for me. Review https://github.com/sdkman/sdkman-cli/issues/83 to see how to enable it for your favorite shell

**Comment by 1pavanb** [comment]
The official `sdk` plugin for zsh is already available in your ohmyzsh installation.
Just enable it in your plugins list `.zshrc` file and restart the terminal.
`plugins=(sdk)`

**Comment by helpermethod** [comment]
Hi!

With Bash completion now in place, we could display the list of installed candidates when autocompleting the `use` command.

That way we wouldn't need to introduce another command only for showing installed version. @marc0der wdyt?

**Comment by mperezi** [comment]
I came up with a pretty geek solution myself by means of a proxy function for the `sdk`command:

```bash
function sdk_proxy {
  if (( $# == 2 )) && [[ $1 =~ ^(u|use)$ ]]; then
    sdk use $candidate $(sdk_fuzzy_find_local_version $2)
  else
    sdk "$@"
  fi
}
```

(The exact mechanism for turning the above into a proxy function in bash is irrelevant here but you can check out https://frederic-hemberger.de/articles/speed-up-initial-zsh-startup-with-lazy-loading/ in case you're interested).

The snippet above just says that whenever I invoke the use subcommand followed by a candidate BUT NOT a version (i.e. `sdk use java` or `sdk u java`) then I plug in a custom fuzzy search:

```bash
function sdk_fuzzy_find_local_version {
  __sdkman_build_version_csv $1 | 
    tr ',' '\n' | 
    sort -r -n | 
    fzf -d '\.' --nth 1
}
```

I know this is a bit fragile as I'm relying on an internal sdkman function (named `__sdkman_build_version_csv`) that is bound to change (or be gone) in the future. But, after all, this is just a workaround and besides I ❤️ proxies  ¯\_(ツ)_/¯

**Comment by marc0der** [comment]
@helpermethod is it worth keeping this issue open until we have a working solution for ZSH too?

**Comment by marc0der** [comment]
@mperezi very cool, and it must have been fun doing this :smile: However, I certainly would recommend that you use the command line completion that @helpermethod just implemented instead. I'll also look at doing the same for ZSH soon.

**Comment by marc0der** [comment]
Re-opening this issue until ZSH is also supported. This will involve porting the oh-my-zsh plugin back into our completion mechanism.

**Comment by helpermethod** [comment]
@marc0der Yes, we should keep the issue open. I can take a stab at ZSH this evening, but I also wouldn't be to sad if anyone else takes over the ZSH completion 😄.

**Comment by marc0der** [comment]
I can have a crack at it sometime, you've already done ample with bash completion.

**Comment by marc0der** [comment]
The autocompletion functionality is now available for both bash and ZSH. It was released as v5.11.1.

**Comment by franco-circleup** [comment]
For zsh I cannot get shell completion enabled from using the instructions:  https://sdkman.io/usage#completion

I tried them as-is as well as replacing `bash` with `zsh`.

**Comment by helpermethod** [comment]
Hi @franco-circleup, it seems an error has sneaked into the docs.

The correct command is

    source <(sdk completion bash)

or for `zsh`

    source <(sdk completion zsh)

I'll fix the documentation tomorrow. Sorry for the inconvenience!

**Comment by franco-circleup** [comment]
That works perfectly, thanks!

**Comment by Abhinav1217** [comment]
I have enabled completion, But I only get completion for first parameter. 

1. I type `sdk `, double tap the tab key.
2. I get list of accepted parameter, I select `use` then  type `java ` 

What is expected
3. double tab should list installed java candidates.
What actually happens. 
3. double tab just list all the folders and files ( this is, AFAIK, the  default  behavior of bash shell )

**Comment by helpermethod** [comment]
Hi @Abhinav1217,

works fine for me. Are you using the `bash` or `zsh` completion?

![image](https://user-images.githubusercontent.com/1562019/118247645-b25c6280-b4a3-11eb-852a-3b967f7b0b75.png)

**Comment by Abhinav1217** [comment]
I am using bash completion, I updated sdk and recreated sdk completion, It did not work. 
Then based on someone's advice on reddit, 

- I removed the line  `source <(sdk completion bash)` 
- I then created a separate completion file by running `sdk completion bash >> 00_sdkman_bash_completion`  in terminal
- I edited the file to remove all the comments, remove broadcast message (if printed) and (optionally) added shebang for bash.
- Sourced that file inside my `~/.bashrc`

Reason given was that sometimes, sdkman prints broadcast messages which will make completion fails. I stumbled upon it as I was occasionally getting this issue whenever I launch new terminal. Which I suspected was something to do with sdk completion but wanted to be sure before I come here. 

![bash_error](https://user-images.githubusercontent.com/740584/118402799-273cc180-b689-11eb-90ca-8e7da087c376.png)

When I created the 00_sdkman_bash_completion file, I did find the broadcast message in the output. I removed it issue was gone, Only downside is that now I would have to manually update the completion file from time to time.  
 
This issue is resolved for me, unless broadcast message should not have been printed when subcommand is completion. In that case, I have no idea why I got it.

**Comment by marc0der** [comment]
Hi @Abhinav1217, we are already aware of this issue and tracking it in #912. We should have a fix out for this soon. In the meanwhile, please source the completion file directly from your profile rather than using the completion command.

**Comment by bentolor** [comment]
Sorry to resurrect this issue that started rather unfortunate.

I'm a `fish` shell user and hence only got to use sdk!man by `omf i sdk`. I stumbled over _exact_ the same issue as OP, and after scrolling through this issue I feel that it is still open and very relevant. **Even more: I still don't know how to achieve my query!**

For illustration. This is what i.e. `omf` tells you after simply typing `omf`:
![grafik](https://user-images.githubusercontent.com/588260/135745384-237816d1-0926-460a-83c1-d0e6bf51bff3.png)
I _immediately_ understand: `omf l` is the way to go.

I was really strained looking at the `sdk` output:
![grafik](https://user-images.githubusercontent.com/588260/135745415-8b68e651-b18d-4534-9d0d-bbcfda443da4.png)

1. I never would have assumed/understood that `sdk current` at least reveals some of the installed candidates
2. The aforementioned `sdk list | grep installed` does not list _any_ hits at all
3. I understood that the rest of this tickets addressed only user-specific shell hack workarounds. Not the think I'd expect from a package manager.

**I'd still propose, that this issue is still relevant and an easy command to the list the locally installed packages should be offered and easily discoverable from the CLI**.

 I feel that should not need to become a shell programming expert or documentation archaeologist to fulfil this task… For me this sounds like a very, very easy and task and undoubtedly relevant functionality. Anybody willing to explain me the backgrounds for this discussion which is ongoing for more than five years?

**Comment by loopmode** [comment]
I agree with @bentolor
What's the expectation of the maintainers? For the cimmunity to create a fork just for the sake of one useful feature that is _very_ very easy to implement without any risk for the existing functionality? (e.g. add an explicit command)

**Comment by pstephenwille** [comment]
still no way to list the installed versions?

**Comment by olledencker** [comment]
I solved it by 
sdk list java|grep installed

**Comment by mikymigs** [comment]
It works but it's workaround... 
Also I have some installs that have the label "local only", so I have to use: `sdk list java | grep -e installed -e local`
This thread if full of workarounds but there's no doubt that some kind of a `sdk list installed` command would be much more user friendly.

**Comment by marc0der** [comment]
Point taken. I'll certainly include a flag listing only installed versions when I rewrite this command.

**Comment by MTyson** [comment]
Guys, sdkman is awesome, it's insane that there isn't a simple in-framework way to say: "What are my installed java versions?" and "let me pick which one to make default" - simply.

**Comment by marc0der** [comment]
@MTyson It's insanely simple!

```
$ sdk default java <tab>
```
Now pick the one you want to make default. Simple, isn't it? :smile:

**Comment by heipl-paul** [comment]
`sdk list java | grep -e installed -e local ` is the actual answer. If you have a version that isn't available from sdkman anymore, it won't show up as installed. Instead it shows up as "Local Only"

That being said, it is both mind blowing and ridiculous that such a command still doesn't exist.

**Comment by marc0der** [comment]
> That being said, it is both mind blowing and ridiculous that such a command still doesn't exist

@heipl-paul this is a free and open source project run by volunteers. We spend our spare time delivering software that you get to use for free every day of your life. No cost and with no strings attached.

If a feature that you really want isn't supported, you can easily get stuck in and contribute by way of a pull request, provided that your code is up to standard. Alternatively, feel free to put down a bounty on our open collective page.

**Comment by mbogner** [comment]
To see what versions are in use:

```shell
➜  ~ sdk current

Using:

gradle: 8.1.1
java: 17.0.7-tem
kotlin: 1.8.20
maven: 3.9.1
visualvm: 2.1.5
```

To list all available and installed versions of tools that at least have one version installed on the machine:

```shell
for i in `ls .sdkman/candidates/`; do sdk list $i; done
```

Haven't found out how to avoid the use of `more`. The suggested grep doesn't work for the multi column syntax used for example for gradle.

Would be cool to have something like `sdk list -h` / `sdk list --help` to get available flags listed.

**Comment by devrimbaris** [comment]
agree with the sdk list --install stuff, would be more useful and intuitive IMHO.
In addition, I have 4 JDK's installed and sometimes I have to use JDK11. Rather than grepping the internals of the sdkman what is the proper way to get  folders for a sdk?

**Comment by davelet** [comment]
Hello, what's the progress? for list installed softwares.

**Comment by cedric-lamalle** [comment]
Hello,
I have created an extension to do the trick. I added a file `sdkman-installed.sh` in `$SDKMAN_DIR/ext` with the following content:

```
#!/usr/bin/env bash

#stolen from src/main/bash/sdkman-list.sh!
function __sdkman_build_version_csv() {
	local candidate versions_csv

	candidate="$1"
	versions_csv=""

	if [[ -d "${SDKMAN_CANDIDATES_DIR}/${candidate}" ]]; then
		for version in $(find "${SDKMAN_CANDIDATES_DIR}/${candidate}" -maxdepth 1 -mindepth 1 \( -type l -o -type d \) -exec basename '{}' \; | sort -r); do
			if [[ "$version" != 'current' ]]; then
				versions_csv="${version},${versions_csv}"
			fi
		done
		versions_csv=${versions_csv%?}
	fi
	echo "$versions_csv"
}

function __sdkman_build_candidate_csv() {
	local candidates_csv
	candidates_csv=""
	if [[ -d "${SDKMAN_CANDIDATES_DIR}" ]]; then
		for candidate in $(find "${SDKMAN_CANDIDATES_DIR}" -maxdepth 1 -mindepth 1 -type d -exec basename '{}' \; | sort -r); do
			candidates_csv="${candidate},${candidates_csv}"
		done
		candidates_csv=${candidates_csv%?}
	fi
	echo "$candidates_csv"
}

function __print_versions_installed() {
	local versions_csv
	versions_csv="$1"
	local versions=($(echo ${versions_csv//,/ }))
	for version in "${versions[@]}"; do
		echo -e "\t* ${version}"
	done
}

function __sdk_installed() {
	local candidates_csv
	candidates_csv=$(__sdkman_build_candidate_csv)
	local candidates=($(echo ${candidates_csv//,/ }))
	for candidate in "${candidates[@]}"; do
			echo "* ${candidate}"
			__print_versions_installed $(__sdkman_build_version_csv "${candidate}")
	done
}
```
Now you can list installed candidates with `sdk installed`.

The code is a little bit rough (there are duplicates...) but if there is interest I can clean it and make some tests to do a PR.

Pro tip: You can format the output with [glow](https://github.com/charmbracelet/glow): `sdk installed | glow -`.

**Comment by bvdriel** [comment]
Hi,

This is an old thread, not sure if in the meantime there is a better solution, but since sdk list will output the 'available' sdk's, when you're in online mode, it will show what is available online, while in offline mode, it shows what is available offline (thus installed).

The solution to your problem using the available commands of sdkman is thus (using example of java here) :
**prompt> sdk offline enable && sdk list java && sdk offline disable**

**Comment by anishp55** [comment]
@bvdriel i tried that, and it doesn't seem to work on MacOS.  i have 11.0.20-tem and 17.0.8-tem installed and it only shows  11.0.20-tem

**Comment by GoldenaArcher** [comment]
I tried something like this:

```bash
❯ ls ~/.sdkman/candidates/* | grep -v "current"
/Users/username/.sdkman/candidates/java:
11.0.16-ms
17.0.4.1-ms
17.0.8.1-ms

/Users/username/.sdkman/candidates/maven:
3.6.3

/Users/username/.sdkman/candidates/tomcat:
```

It's pretty similar to what [mbogner](https://github.com/mbogner) has done in the previous thread.

**Comment by mjustin** [comment]
> `sdk list java | grep -e installed -e local ` is the actual answer. If you have a version that isn't available from sdkman anymore, it won't show up as installed. Instead it shows up as "Local Only"

This doesn't show my older `hs-adpt` Java installs.

```
$ sdk use java 
11.0.9.hs-adpt   13.0.2.hs-adpt   14.0.2.hs-adpt   15.0.1.hs-adpt   15.0.2.hs-adpt   16.0.1.hs-adpt   17.0.0-tem       19.0.2-tem       8.0.275.hs-adpt
```

```
$ sdk list java | grep -e installed -e local
We periodically need to update the local cache. Please run:
               |     | 19.0.2       | tem     | local only | 19.0.2-tem          
               | >>> | 17.0.0       | tem     | local only | 17.0.0-tem    
```

**Comment by dominictobias** [comment]
Appreciate this is free, but it's still odd to make an SDK manager without a simple command to list your installed versions and which is currently selected. That would be like the first command I'd think of adding. Any other manager I've used has this feature like `nvm list`, `rbenv versions`, `pyenv versions`, `phpenv versions` etc

**Comment by sdavids** [comment]
Could we maybe reopen this issue? — A lot of interest by the community …

**Comment by mbogner** [comment]
Hi @sdavids,
did you see the comment with `sdk current`? From my perspective this solves it for me. There you can see all installed packages (via sdkman) and the version. If you then want to drill down what versions exist for something you can do `sdk list <name>`. So you have the overview and details call. What are you missing exactly?

**Comment by mbogner** [comment]
- `sdk list`

```

Using:

gradle: 8.7
java: 21.0.2-graalce
kotlin: 1.9.23
maven: 3.9.6
visualvm: 2.1.8
```

- `sdk list maven`

```
================================================================================
Available Maven Versions
================================================================================
     4.0.0-alpha-13      3.8.8               3.5.0
     4.0.0-alpha-12      3.8.7               3.3.9
     4.0.0-alpha-10      3.8.6               3.3.3
     4.0.0-alpha-9       3.8.5               3.3.1
     4.0.0-alpha-8       3.8.4               3.2.5
     4.0.0-alpha-7       3.8.3               3.2.3
     4.0.0-alpha-5       3.8.2               3.2.2
     4.0.0-alpha-4       3.8.1               3.2.1
 > * 3.9.6               3.6.3               3.1.1
     3.9.5               3.6.2               3.1.0
     3.9.4               3.6.1               3.0.5
     3.9.3               3.6.0               3.0.4
     3.9.2               3.5.4
     3.9.1               3.5.3
     3.9.0               3.5.2

================================================================================
+ - local version
* - installed
> - currently in use
================================================================================
```

- `sdk list`

```
================================================================================
Available Candidates
================================================================================
q-quit                                  /-search down
j-down                                  ?-search up
k-up                                    h-help

--------------------------------------------------------------------------------
Apache ActiveMQ (Classic) (5.17.1)                  https://activemq.apache.org/

Apache ActiveMQ® is a popular open source, multi-protocol, Java-based message
broker. It supports industry standard protocols so users get the benefits of
client choices across a broad range of languages and platforms. Connect from
clients written in JavaScript, C, C++, Python, .Net, and more. Integrate your
multi-platform applications using the ubiquitous AMQP protocol. Exchange
messages between your web applications using STOMP over websockets. Manage your
IoT devices using MQTT. Support your existing JMS infrastructure and beyond.
ActiveMQ offers the power and flexibility to support any messaging use-case.

                                                          $ sdk install activemq
--------------------------------------------------------------------------------
Ant (1.10.13)                                            https://ant.apache.org/

Apache Ant is a Java library and command-line tool whose mission is to drive
processes described in build files as targets and extension points dependent
upon each other. The main known usage of Ant is the build of Java applications.
Ant supplies a number of built-in tasks allowing to compile, assemble, test and
run Java applications. Ant can also be used effectively to build non Java
applications, for instance C or C++ applications. More generally, Ant can be
used to pilot any type of process which can be described in terms of targets and
tasks.

                                                               $ sdk install ant
--------------------------------------------------------------------------------
AsciidoctorJ (2.5.12)                                    http://asciidoctor.org/

AsciidoctorJ is the official library for running Asciidoctor on the JVM. Using
AsciidoctorJ, you can convert AsciiDoc content or analyze the structure of a
parsed AsciiDoc document from Java and other JVM languages.

                                                      $ sdk install asciidoctorj
--------------------------------------------------------------------------------
Ballerina (swan-lake-p3)                                   https://ballerina.io/

Open source programming language and platform for cloud-era application
...
```
```

**Comment by sdavids** [comment]
@mbogner 

I would like to have something similar to 

```shell
$ apt list --installed
```

> list
>      list [...] can display [...] installed (--installed) [...] versions.

https://manpages.ubuntu.com/manpages/xenial/man8/apt.8.html#description

or 

```shell
$ dnf list --installed
```

> **dnf [options] list --installed [<package-file-spec>...]**
>              Lists installed packages.

https://www.man7.org/linux/man-pages/man8/dnf.8.html

or 

```shell
$ brew list
```

> ### list
>List all installed formulae.

https://docs.brew.sh/Manpage#list

## Status Quo

```shell
$ sdk current

Using:

ant: 1.10.13
gradle: 8.7
java: 21.0.2-tem
maven: 3.9.6
$ sdk offline enable && sdk list java && sdk offline disable
Offline mode enabled.
--------------------------------------------------------------------------------
Offline: only showing installed java versions
--------------------------------------------------------------------------------
 * 22-tem
 > 21.0.2-tem
 * 17.0.10-tem
 * 11.0.22-tem
--------------------------------------------------------------------------------
* - installed
> - currently in use
--------------------------------------------------------------------------------
Online mode re-enabled!
$ sdk list java | grep -e installed -e local
               |     | 8.0.402      | librca  | installed  | 8.0.402-librca
 Temurin       |     | 22           | tem     | installed  | 22-tem
               | >>> | 21.0.2       | tem     | installed  | 21.0.2-tem
               |     | 17.0.10      | tem     | installed  | 17.0.10-tem
               |     | 11.0.22      | tem     | installed  | 11.0.22-tem
```
## Feature Request

```shell
$ sdk list java --installed
================================================================================
Installed Java Versions
================================================================================
 Vendor        | Use | Version      | Dist    | Identifier
--------------------------------------------------------------------------------      
 Liberica      |     | 8.0.402      | librca  | 8.0.402-librca  
 Temurin       |     | 22           | tem     | 22-tem              
               | >>> | 21.0.2       | tem     | 21.0.2-tem             
               |     | 17.0.10      | tem     | 17.0.10-tem            
               |     | 11.0.22      | tem     | 11.0.22-tem         
```

**Comment by mbogner** [comment]
@sdavids Ok, got your point.

Isn't 

```shell
apt list --installed
```

```
Listing... Done
adduser/jammy,now 3.118ubuntu5 all [installed,automatic]
amd64-microcode/jammy-updates,jammy-security,now 3.20191218.1ubuntu2.2 amd64 [installed,automatic]
apache2-bin/jammy-updates,now 2.4.52-1ubuntu4.8 amd64 [installed,automatic]
apache2-data/jammy-updates,now 2.4.52-1ubuntu4.8 all [installed,automatic]
apache2-utils/jammy-updates,now 2.4.52-1ubuntu4.8 amd64 [installed,automatic]
apache2/jammy-updates,now 2.4.52-1ubuntu4.8 amd64 [installed]
apparmor/jammy-updates,now 3.0.4-2ubuntu2.3 amd64 [installed,automatic]
apport-symptoms/jammy,now 0.24 all [installed,automatic]
apport/jammy-updates,now 2.20.11-0ubuntu82.5 all [installed,automatic]
apt-utils/jammy-updates,now 2.4.11 amd64 [installed,automatic]
apt/jammy-updates,now 2.4.11 amd64 [installed,automatic]
augeas-lenses/jammy,now 1.13.0-1 all [installed,automatic]
base-files/jammy-updates,now 12ubuntu4.6 amd64 [installed]
...
```

the same as 

```shell
sdk current
```
?

Afaik you can't have multiple versions of the same debian package installed at the same time. That would be the selected one with sdkman. I agree that it would be nice for maintenance of a system to have a command that just lists all locally installed packages instead of just the currently selected version. Even more handy for me would be to have an option to uninstall all currently not used versions of all packages (or of listed packages). Something like this:

```
sdk uninstall java --keep-current
```

or for all installed packages ->

```
sdk uninstall --keep-current
```

**Comment by sdavids** [comment]
I guess it is a question of how one defines "installed" …

a) "exists on the file system"
b) "accessible via $PATH"

**Comment by ikaronen-relex** [comment]
I'm using SDKMan to juggle between different major Java versions needed to build, test and benchmark different LTS branches of our software. The `sdk installed` feature [requested above](https://github.com/sdkman/sdkman-cli/issues/466#issuecomment-2025237789) by @sdavids sounds exactly like what I need.

I'm aware that `sdk list java | egrep 'installed|local'` already does more or less what I want, but it's quite an awkward command to use.

**Comment by fedeValtira** [comment]
why is closed?

**Comment by troyere** [comment]
So there is no such command even in 2024 ? 🤣

**Comment by mbogner** [comment]
> So there is no such command even in 2024 ? 🤣

no. and it was closed. just workarounds

**Comment by joaquincorimayo** [comment]
So... I continue using grep for list installed?

**Comment by mbogner** [comment]
> So... I continue using grep for list installed?

see my answer from Mar 28. sdk current for listing active versions and sdk list <x> for available versions. Fine for me.

**Comment by oviniciusfeitosa** [comment]
- List installed java versions by SDKMAN: `ls -la ~/.sdkman/candidates/java`
- List installed maven versions by SDKMAN: `ls -la ~/.sdkman/candidates/maven`
- List installed springboot versions by SDKMAN: `ls -la ~/.sdkman/candidates/springboot`

**Comment by apotheque** [comment]
this issue is a monument to man's arrogance

**Comment by marc0der** [comment]
@apotheque and clearly you haven't re-read the entire thread.

**Comment by marc0der** [comment]
**TLDR;** Use the tab key to see all installed versions. When this command is rewritten in Rust, we will implement a special `--installed` flag.

**Comment by apotheque** [comment]
> [@apotheque](https://github.com/apotheque) and clearly you haven't read the entire thread.

sorry, I didn't see all the autocompletion discussion as it was folded. really happy to hear you're going to bring us the --installed flag 😊

**Comment by fResult** [comment]
I found an alternative way to list installed SDKs using `sdk` commands.

**First, see all your candidates using `sdk current`:**

```console
➜ sdk current

Using:

gradle: 8.10.2
java: 23.0.1-librca
sbt: 1.7.1
scala: 3.3.3
```

This shows I have `gradle`, `java`, `sbt`, and `scala` installed.

**List installed Java SDKs using `sdk list java | grep installed`:**

```console
➜ sdk list java | grep installed
 GraalVM CE    |     | 23.0.2       | graalce | installed  | 23.0.2-graalce      
               | >>> | 23.0.1       | librca  | installed  | 23.0.1-librca       
               |     | 21.0.5       | librca  | installed  | 21.0.5-librca
```

This shows my installed Java versions are `23.0.2:graalce`, `23.0.1:librca`, and `21.0.5:librca`.

**List installed SBT SDKs using `sdk list sbt | grep \*`:**

```console
➜ sdk list sbt | grep \*
   * 1.10.7              1.8.0               1.4.5               1.2.6          
     1.10.4          > * 1.7.1               1.4.2               1.2.3          
* - installed
```

This shows my installed SBT versions are `1.10.7` and `1.7.1`.

For more details, check out: https://medium.com/p/f809d7891f1d

**Comment by russeg** [comment]
grepping is good enough i guess. you would think listing installed versions is one of the very first command to implement at the very beginning of the project, but no lol.

also maintainers (of any project not just this one) asking for pr is pretty dumb, since most prs are not merged and mostly ignored.
